test/test_fork.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2022-present MongoDB, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Test that pymongo is fork safe."""
+
+import os
+import threading
+from multiprocessing import Pipe
+from test import IntegrationTest, client_context
+from typing import Any, Callable
+from unittest import skipIf
+from unittest.mock import patch
+
+from bson.objectid import ObjectId
+from pymongo import MongoClient
+from pymongo.lock import _ForkLock
+
+
+@client_context.require_connection
+def setUpModule():
+    pass
+
+
+class ForkThread(threading.Thread):
+    def __init__(self, exit_cond: Callable[[], int] = lambda: 0):
+        super().__init__()
+        self.pid: int = -1  # Indicate we haven't started.
+        self.exit_cond = exit_cond
+
+    def run(self):
+        if self.pid < 0:
+            self.pid = os.fork()
+            if self.pid == 0:
+                os._exit(self.exit_cond())
+
+
+class LockWrapper:
+    def __init__(self, fork_thread: ForkThread, lock_type: Any = _ForkLock):
+        self.__lock = lock_type()
+        self.fork_thread = fork_thread
+
+    def __enter__(self):
+        self.__lock.__enter__()
+        if not self.fork_thread.is_alive():
+            self.fork_thread.start()
```

Comment:
start() is non-blocking. The thread might not actually start (and finally call fork()) until much later. Is that expected or did you want to block here until fork() was called?

Reply:
Yeah I think I assumed it would have run before leaving the critical region, I added a join in the `__exit__` wrapper to make sure it runs all the way in the critical region.